### PR TITLE
Describe the core concepts in the README file.

### PR DESCRIPTION
**Pull request by jmalloc**
~This is a PR under which I will re-address all documentation.~ (now just a part of #60). It's based off the `43-external-interfaces` branch, but I will rebase it once PR #57 has been merged. (This is done now).

I'm putting it here now for visibility, but I wont add you guys (@danilvpetrov, @koden-km) as reviewers until I am think it is ready to be merged.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,5 +7,142 @@
...
+
+Dogma defines four handler types, one each for [aggregates](#Aggregate),
+[processess](#Process), [integrations](#Integration) and
+[projections](#Projection). These concepts are described in more detail
+below.
+
+### Application
+
+An **application** is a set of message handlers that operate together as a unit.
+Applications are represented by the [`dogma.App`] struct.
```
Comment (koden-km) [inline README.md]: This is now a `dogma.Application` interface?

File: README.md [hunk 2]
```diff
@@ -7,5 +7,142 @@
...
+
+Integrations are represented by the [`dogma.IntegrationMessageHandler`] interface.
+
+### Projection
+
+A **projection** is a unit of application logic that derives some specific
+portion of application state from the events that occur. This state is often
+referred to as a "read model".
+
+Projections recieve event messages and do not produce messages of any kind.
```
Comment (koden-km) [inline README.md]:
```suggestion
Projections receive event messages and do not produce messages of any kind.
```

**Comment by koden-km** [comment]
Ah, maybe i jumped in to review this too early. It had a bunch of the PR's open and was just going through them.

**Comment by jmalloc** [comment]
@koden-km FYI, this one is here now for discussion, but needs a lot more work before it's ready to commit.

**Comment by jmalloc** [comment]
Some notes from @ezzatron, regarding the README:

---

>> This module does not provide any engine implementation, but rather defines the API to be implemented by both application developers and engine developers.

> Do application developers implement it? Or utilise it?

I mentioned that this is documented on each interface in the API docs, but it might be good to mention that here.

---

>> An aggregate is a unit of application logic with associated state that encodes the business invariants of a specific application. The concept is taken directly from Domain Driven Design.

> "business invariants" is probably an uncommon enough term to benefit from a definition link
> or if it can be worded using simpler terms?

---

>>Aggregate state is managed by the engine, ensuring that state transitions occur atomically with the recording of the events about those state transitions.

> This is one of the key features of Dogma, right? Might deserve some more emphasis

---

> it reads a bit like a glossary
> I think some heavier cross-linking between the terms would be good
> even for things like message and command
> minor readability criticism: there's a lot of single-sentence paragraphs
> also I still struggle a little bit in differentiating aggregates and processes, that might be a problem for > fresh readers
> I do get the long-running aspect though

**Comment by jmalloc** [comment]
I would like to merge this in rather than keeping it open until all docs are done, I've created #60 to track the broader task of rewording the documentation.

File: README.md [hunk 3]
```diff
@@ -7,5 +7,142 @@
...
+Integrations receive command messages and produce event messages. They do not
+have any state that is managed by the engine.
+
+Integrations are represented by the [`dogma.IntegrationMessageHandler`] interface.
+
+### Projection
+
+A **projection** is a unit of application logic that derives some specific
+portion of application state from the events that occur. This state is often
+referred to as a "read model".
```
[the remark this answers is not in the document]
Comment (jmalloc) [reply README.md]: Are they (often) ?  I actually didn't think that was super common, but I know you found something in the big book yesterday.

[the remark this answers is not in the document]
Comment (koden-km) [reply README.md]: Hmm ok, i thought that was their one purpose.

[the remark this answers is not in the document]
Comment (koden-km) [reply README.md]: This is from the DDD book page 140:
> Known by Other Names
> Note that some areas/components of CQRS may be known by other names. What I call the query model is also known as the read model, and the command model is also called the write model.

File: README.md [hunk 4]
```diff
@@ -27,9 +27,10 @@ various design patterns such as [CQRS], [Event Sourcing] and [Eventual Consisten
 ### Message and Message Handler
 
 A **message** is an application-defined unit of data that encapsulates a
-**command** or **event** within a message-based application. Command messages
-represent a request for the application to perform some action, whereas event
-messages indicate that some action has already occurred.
+**command** or **event** within a message-based application. A command messages
```
Comment (koden-km) [inline README.md]:
```suggestion
**command** or **event** within a message-based application. A command message
```

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (danilvpetrov) [reply README.md]: Hmm, what is `command model` then? Is this an aggregate?

File: README.md [hunk 5]
```diff
@@ -48,39 +49,49 @@ Command messages are always routed to exactly one handler. Event messages may
...
+changes are represented by event messages. By definition, changes to the state
+of an aggregate instance are ["immediately consistent"][Immediate Consistency]
+(aka "transactionally consistent"). This means that the results of a command
+against a given instance are always visible to subsequent commands for that
+instance.
+
+Aggregate state is managed by the [engine](#engine), ensuring that state
+transitions occur atomically with the recording of the events that describe
+those state
+transitions.
```
Comment (koden-km) [inline README.md]: There is a extra line break here.

File: README.md [hunk 6]
```diff
@@ -7,5 +7,164 @@
...
+the same process instance at a specific time. Such messages are used to
+implement processes that incorporate some time-based component.
+
+Because a process coordinates changes within the application using multiple
+messages, and each message represents a single atomic change to the application's
+state, the changes made by a process are ["eventually consistent"][Eventual Consistency].
+The precise guarantees regarding process consistency are specific to the [engine](#Engine) implementation.
+
+Process state is managed by the engine, ensuring that state transitions occur
+atomically with the enqueing of the commands the processes produce.
```
Comment (koden-km) [inline README.md]:
```suggestion
atomically with the enqueing of the commands that the processes produce.
```

Comment (jmalloc) [reply README.md]: I propose rewording this whole paragraph:

```
Aggregate state is managed by the [engine](#engine), ensuring that changes to a
specific instance and the recording of events that represent those changes occur
atomically.
```

Comment (danilvpetrov) [reply README.md]: that sounds clear to me

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmalloc) [reply README.md]: Yep, I would say under that CQRS terminology an aggregate is definitely the "command model", however there's nothing in Dogma that demands CQRS.  The aggregate state could be represented by regular database tables which are both the read and write models.

Dogma projections, on the other hand, are always read models.

[the remark this answers is not in the document]
Comment (danilvpetrov) [reply README.md]: Ok, that makes sense, thanks.

[the remark this answers is not in the document]
Comment (jmalloc) [reply README.md]: We could say something along the lines of 

> When using CQRS, the aggregate is the "write model" or "command model".

We **do** already make mention of CQRS in the introductory paragraphs.

[the remark this answers is not in the document]
Comment (koden-km) [reply README.md]: I like your suggested line, but if you want to leave it out that's fine.

[the remark this answers is not in the document]
Comment (jmalloc) [reply README.md]: Ok, before I commit it, how about these change:

```diff
diff --git a/README.md b/README.md
index 6bbeb37..ee64c2c 100644
--- a/README.md
+++ b/README.md
@@ -78,7 +78,10 @@ engine or the application.
 An **aggregate** is a unit of application logic with associated state that
 encodes the business invariants of a specific application. These invariants are
 the "rules" of the business domain that must not be violated, even temporarily.
-This concept is taken directly from [Domain Driven Design].
+
+The aggregate concept is taken directly from [Domain Driven Design]. When
+employing [CQRS], the aggregate is sometimes referred to as the "write model",
+or the "command model".
 
 An aggregate receives command [messages](#message-and-message-handler) in order
 to effect a change in a particular **instance** of that aggregate. Such state
@@ -139,7 +142,8 @@ Integrations are represented by the [`dogma.IntegrationMessageHandler`] interfac
 
 A **projection** is a unit of application logic that derives some specific
 portion of application state from the events that occur. This state is often
-referred to as a "read model".
+referred to as a "read model" or "query model", especially when employing the
+[CQRS] pattern.
 
 Projections receive event [messages](#message-and-message-handler) and do not
 produce messages of any kind.
```

[the remark this answers is not in the document]
Comment (jmalloc) [reply README.md]: Actually, let me tweak that a bit.

[the remark this answers is not in the document]
Comment (jmalloc) [reply README.md]: Ok, https://github.com/dogmatiq/dogma/pull/58#discussion_r updated.

[the remark this answers is not in the document]
Comment (koden-km) [reply README.md]: Looks good, ship it.

[the remark this answers is not in the document]
Comment (jmalloc) [reply README.md]: Ok, click approve you heathens :) so I can feel good about myself.

[the remark this answers is not in the document]
Comment (koden-km) [reply README.md]: I can now that you pushed your diff ;)

**Pull request merged**